Save memory by moving more RenderStyle Vectors to FixedVector

weinig opened the pull request:
#### 7a8e21f8b0bd5d0377eb851f7f090086c34ad473
<pre>
Save memory by moving more RenderStyle Vectors to FixedVector
<a href="https://bugs.webkit.org/show_bug.cgi?id=293855">https://bugs.webkit.org/show_bug.cgi?id=293855</a>

Reviewed by Darin Adler.

Moves an initial set of RenderStyle Vectors to FixedVector.

* Source/WTF/wtf/EmbeddedFixedVector.h:
* Source/WTF/wtf/FixedVector.h:
* Source/WTF/wtf/TrailingArray.h:
    - Add support for construction via std::initializer_list, std::span and
      mapped sized range.
    - Add search functions matching those in Vector.

* Source/JavaScriptCore/bytecode/InlineCacheCompiler.cpp:
    - Replace use of createWithSizeFromGenerator with direct std::initializer_list use.

* Source/WTF/wtf/text/TextStream.h:
    - Match Vector, defining supports_text_stream_insertion for FixedVector.

* Source/WebCore/css/CSSProperties.json:
    - Remove a bunch of now unnecessary explicit template parameters to interpolation wrapper types.

* Source/WebCore/css/scripts/process-css-properties.py:
    - Add missing to forwardInheritedValue for color properties (needed now due to setters taking
      r-values instead of const references).

* Source/WebCore/dom/ViewTransition.cpp:
    - Update to avoid copy now that viewTransitionName() returns a const reference.

* Source/WebCore/rendering/style/RenderStyle.cpp:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/rendering/style/RenderStyleSetters.h:
* Source/WebCore/rendering/style/SVGRenderStyle.h:
* Source/WebCore/rendering/style/SVGRenderStyleDefs.h:
* Source/WebCore/rendering/style/StyleGridData.h:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
* Source/WebCore/rendering/svg/SVGRenderSupport.cpp:
* Source/WebCore/rendering/svg/SVGRenderTreeAsText.cpp:
* Source/WebCore/style/StyleBuilderConverter.h:
* Source/WebCore/style/StyleBuilderCustom.h:
* Source/WebCore/style/StyleExtractorConverter.h:
* Source/WebCore/style/StyleExtractorSerializer.h:
* Source/WebCore/style/StyleInterpolationFunctions.h:
* Source/WebCore/style/StyleInterpolationWrappers.h:
    - Update setters/getters to use FixedVector instead of Vector.
    - Update setters/getters to return const reference and take r-value reference
      for non-trivial types.

* Source/WebCore/platform/graphics/DashArray.h:
* Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp:
* Source/WebCore/platform/mock/MockRealtimeVideoSource.cpp:
* Source/WebCore/rendering/BorderPainter.cpp:
    - Switch DashArray to use a FixedVector.

* Source/WebKit/Platform/IPC/ArgumentCoders.h:
    - Add coder for FixedVector.

* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
    - Switch DashArray to FixedVector definition.

Canonical link: <a href="https://commits.webkit.org/295663@main">https://commits.webkit.org/295663@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5f4e962af9d401b5187fecea35feecc91530d6dd

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/105842 "3 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/25593 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/15986 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/111039 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/56439 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/107883 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/26175 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/34096 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/80391 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/108848 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/20570 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/95511 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/60706 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/20233 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/13608 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/55877 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/98483 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/89974 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/13646 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/113889 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/104461 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/32982 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/24313 "Found 3 new test failures: imported/w3c/web-platform-tests/mimesniff/media/media-sniff.window.html platform/mac/fast/text/line-break-locale.html svg/zoom/page/text-with-non-scaling-stroke.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/89469 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/33346 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/91741 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/89140 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22714 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/34008 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/11815 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/28499 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/32907 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/38318 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/128773 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/32653 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/35148 "Found 1 new JSC stress test failure: wasm.yaml/wasm/gc/ref-i31-eq.js.wasm-eager-jettison (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/36002 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/34251 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WTF/wtf/FixedVector.h:
```diff
@@ -190,9 +196,12 @@ class FixedVector {
+    size_t find(const auto&) const;
+    size_t findIf(NOESCAPE const Invocable<bool(const T&)> auto&) const;
+    size_t reverseFind(const auto&) const;
+    size_t reverseFindIf(NOESCAPE const Invocable<bool(const T&)> auto& matches) const;
```
```suggestion
    size_t reverseFindIf(NOESCAPE const Invocable<bool(const T&)> auto&) const;
```

darinadler commented on Source/WTF/wtf/TrailingArray.h:
```diff
@@ -27,6 +27,7 @@
 
 #include <concepts>
 #include <type_traits>
+#include <wtf/IndexedRange.h>
```
I don't think this is needed to compile the function template that takes `SizedRange`.

weinig replied:
Oops. I tried to use this in an earlier version ended up not being worth it. Will remove.

darinadler commented on Source/WebCore/rendering/style/StyleGridData.h:
```diff
@@ -29,6 +29,7 @@
 #include "GridTrackSize.h"
 #include "RenderStyleConstants.h"
 #include "StyleContentAlignmentData.h"
+#include <wtf/FixedVector.h>
```
Surprising to need this but no other changes in this header. Related: Any chance we can remove more `Vector.h` includes?

darinadler commented on Source/WebKit/Platform/IPC/ArgumentCoders.h:
```diff
@@ -510,6 +510,67 @@ template<typename T, size_t inlineCapacity, typename OverflowHandler, size_t min
+        if (!size)
+            return std::nullopt;
+
+        // Limit direct allocations from untrusted sources to 1MB.
```
It's conservative and safe to match `Vector` in this respect but I suspect not helpful. Is there a chance the `Vector` issue has to do with a specific problem that doesn't exist here?

weinig replied:
I don't think there is any difference. reserveInitialCapacity vs. FixedVector should be the same from a risk factor.

weinig committed:
Save memory by moving more RenderStyle Vectors to FixedVector
https://bugs.webkit.org/show_bug.cgi?id=293855

Reviewed by Darin Adler.

Moves an initial set of RenderStyle Vectors to FixedVector.

* Source/WTF/wtf/EmbeddedFixedVector.h:
* Source/WTF/wtf/FixedVector.h:
* Source/WTF/wtf/TrailingArray.h:
    - Add support for construction via std::initializer_list, std::span and
      mapped sized range.
    - Add search functions matching those in Vector.

* Source/JavaScriptCore/bytecode/InlineCacheCompiler.cpp:
    - Replace use of createWithSizeFromGenerator with direct std::initializer_list use.

* Source/WTF/wtf/text/TextStream.h:
    - Match Vector, defining supports_text_stream_insertion for FixedVector.

* Source/WebCore/css/CSSProperties.json:
    - Remove a bunch of now unnecessary explicit template parameters to interpolation wrapper types.

* Source/WebCore/css/scripts/process-css-properties.py:
    - Add missing to forwardInheritedValue for color properties (needed now due to setters taking
      r-values instead of const references).

* Source/WebCore/dom/ViewTransition.cpp:
    - Update to avoid copy now that viewTransitionName() returns a const reference.

* Source/WebCore/rendering/style/RenderStyle.cpp:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/rendering/style/RenderStyleSetters.h:
* Source/WebCore/rendering/style/SVGRenderStyle.h:
* Source/WebCore/rendering/style/SVGRenderStyleDefs.h:
* Source/WebCore/rendering/style/StyleGridData.h:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
* Source/WebCore/rendering/svg/SVGRenderSupport.cpp:
* Source/WebCore/rendering/svg/SVGRenderTreeAsText.cpp:
* Source/WebCore/style/StyleBuilderConverter.h:
* Source/WebCore/style/StyleBuilderCustom.h:
* Source/WebCore/style/StyleExtractorConverter.h:
* Source/WebCore/style/StyleExtractorSerializer.h:
* Source/WebCore/style/StyleInterpolationFunctions.h:
* Source/WebCore/style/StyleInterpolationWrappers.h:
    - Update setters/getters to use FixedVector instead of Vector.
    - Update setters/getters to return const reference and take r-value reference
      for non-trivial types.

* Source/WebCore/platform/graphics/DashArray.h:
* Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp:
* Source/WebCore/platform/mock/MockRealtimeVideoSource.cpp:
* Source/WebCore/rendering/BorderPainter.cpp:
    - Switch DashArray to use a FixedVector.

* Source/WebKit/Platform/IPC/ArgumentCoders.h:
    - Add coder for FixedVector.

* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
    - Switch DashArray to FixedVector definition.

Canonical link: https://commits.webkit.org/295663@main
